### PR TITLE
Implement render_template & render_partial subscriptions

### PR DESCRIPTION
**Pull request by ChrisBr**
Initial proposal to implement subscriptions for :eyes: 

* render_template.action_view
* render_partial.action_view

https://guides.rubyonrails.org/active_support_instrumentation.html#action-view

* Implemented ``ActionView::RenderSubscription`` class to handle the events and write to influxdb
* Extracted convert_timestamp method to a utils class

I would also propose to move ``InfluxDB::Rails.handle_action_controller_metrics`` and the lambda in ``Railties`` class to a dedicated class.

Thanks for providing feedback :heart:

### CODE REVIEW

**Comment by ChrisBr** [comment]
Ok, added now some unit and integration tests.

**Comment by dmke** [comment]
Sorry, I'm currently a bit busy. Give me a few days to get things in order.

At a first glance this looks promising. I'm not a big fan of `Util` classes, as they tend to end up cluttered with unrelated methods. For now, this is OK, but I'm tempted to move it to`influxdb-ruby` itself (as I can see the usefulness for other consumers).

> I would also propose to move `InfluxDB::Rails.handle_action_controller_metrics` and the lambda in `Railties` class to a dedicated class.

If that clears things up, go ahead.

**Comment by ChrisBr** [comment]
> Sorry, I'm currently a bit busy. Give me a few days to get things in order.

No worries, take your time :watch: 

> At a first glance this looks promising. I'm not a big fan of Util classes, as they tend to end up cluttered with unrelated methods. For now, this is OK, but I'm tempted to move it to influxdb-ruby itself (as I can see the usefulness for other consumers).

Right. Maybe we can move it to a ``TimeConversion`` module to limit the scope of the mixin for now? 

> If that clears things up, go ahead.

Ok, cool! I will try to move it to a dedicated class then.

**Comment by dmke** [comment]
FYI: I've relocated the `convert_timestamp` method to the `influxdb` gem (v0.6.2, [`InfluxDB.convert_timestamp`](https://github.com/influxdata/influxdb-ruby/blob/master/lib/influxdb/timestamp_conversion.rb)), and bumped the dependency here.

**Comment by ChrisBr** [comment]
> FYI: I've relocated the convert_timestamp method to the influxdb gem (v0.6.2, InfluxDB.convert_timestamp), and bumped the dependency here.

:+1: cool, I will update tomorrow.

I added two more features:

* 6517dc3021f6f59a0732dd66265fdc4a6d9408de: Adding the location to the view renderer to track down in which controller action the view was renderer

* 1042b80729d61e1dd7035d05a488b7c5baf38bf9: Tracking of ActiveRecord SQL, the SQL gets normalized before inserted as tags

Thanks

**Comment by dmke** [comment]
> * 6517dc3021f6f59a0732dd66265fdc4a6d9408de: Adding the location to the view renderer to track down in which controller action the view was renderer

Neat :-)

> * 1042b80729d61e1dd7035d05a488b7c5baf38bf9: Tracking of ActiveRecord SQL, the SQL gets normalized before inserted as tags

After reverting `payload[:path]` (#54), I wonder if this wouldn't have the same effect... I would rather have the normalized query as *field value*, and use table name and operation as tags (i.e. `select * from users where id = xxx` would yield `tags: { table_name: "users", op: "select" }, values: { query: "select * ..." }`). What do you think?

**Comment by ChrisBr** [comment]
@dmke I rebased now on the tag middleware changes and also removed the timestamp conversion module.

> After reverting payload[:path] (#54), I wonder if this wouldn't have the same effect... I would rather have the normalized query as field value, and use table name and operation as tags (i.e. select * from users where id = xxx would yield tags: { table_name: "users", op: "select" }, values: { query: "select * ..." }). What do you think?

I guess / hope not as I only track ``SELECT INSERT UPDATE DELETE`` and they get normalized from their parameters so I guess the number of different queries will be limited. But I can try it out for a few days and check what happens...

I just would like to do queries like:
```
SELECT value FROM rails.sql GROUP_BY sql;
```

to see the slowest queries. I'm not too familiar with InfluxDB yet but would that be possible with a field value or what would be the advantage of the field value except that it is not limited in the number of different values like a tag?).

I can also move this last commit to a dedicated PR so that we can get in the PR and continue discussion in the other PR?

**Comment by ChrisBr** [comment]
I would like to get something like this:

selection_013

[Skylight](https://www.skylight.io) does something similar, they combine it with the partial and view results:

selection_014

Would that be possible with a field value? Otherwise maybe we can normalize the queries even more? But I really see a huge advantage in measuring this data :mag:

**Comment by dmke** [comment]
> I guess / hope not as I only track SELECT INSERT UPDATE DELETE and they get normalized from their parameters so I guess the number of different queries will be limited. But I can try it out for a few days and check what happens...

I'll sample a few queries from a few different Rails apps I've access to, and throw it against your normalizer. I suspect the interesting queries (those with sub-sejects, joins, and/or group-by clauses) are going to be boiled down too much. I fear that the simple regex is both too simple has a too large CPU footprint (I've seen hundrets of queries on a single page).

> I just would like [...] to see the slowest queries. I'm not too familiar with InfluxDB yet but would that be possible with a field value or what would be the advantage of the field value except that it is not limited in the number of different values like a tag?).

Tags are automatically indexed and allow to be `GROUP BY`'ed on the server. Fields may hold arbitrary data, but you cannot use field keys in `GROUP BY` clauses. You may want to `GROUP BY sql`, if you need to sample the average (or other percentiles) for analysis.

Say you want to see the average query time in the last 24 hours in 1 hour slices, for each query. This is what you'd send to the server:

```sql
select mean(value) from "rails.sql" where time > now() - 1d group by time(1h), sql
```

If `sql` is a field, this would not be possible.

> Skylight does something similar, they combine it with the partial and view results

That looks to me, as if the omit the conditions from the query (I see `select from venues`, but no `select from venue where x in (y)`. This suggests two tags (`table_name` and `op`).

> I can also move this last commit to a dedicated PR so that we can get in the PR and continue discussion in the other PR?

Yeah, let's do that for now.

**Comment by ChrisBr** [comment]
>> I can also move this last commit to a dedicated PR so that we can get in the PR and continue discussion in the other PR?

> Yeah, let's do that for now.

Dropped and opened https://github.com/influxdata/influxdb-rails/pull/55

File: lib/influxdb/rails/middleware/render_subscriber.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+module InfluxDB
+  module Rails
+    module Middleware
+      class RenderSubscriber
+        include InfluxDB::Rails::Logger
+
+        attr_reader :series_name, :config
+
+        def initialize(config, series_name)
+          @config = config
```
Comment (dmke) [inline lib/influxdb/rails/middleware/render_subscriber.rb]: Can you move the `config` to a base class (e.g. `InfluxDB::Rails::Middleware::Subscriber`), common to both this class and `RequestSubscriber`?

File: lib/influxdb/rails/middleware/render_subscriber.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          @series_name = series_name
+        end
+
+        def call(_name, started, finished, _unique_id, payload)
+          return if !config.instrumentation_enabled? ||
+                    config.ignore_current_environment?
```
Comment (dmke) [inline lib/influxdb/rails/middleware/render_subscriber.rb]: Please refactor this into `return unless enabled?` (needs an `enabled?` method in the base class).

File: lib/influxdb/rails/middleware/request_subscriber.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+require "influxdb/rails/logger"
+
+module InfluxDB
+  module Rails
+    module Middleware
+      class RequestSubscriber
+        include InfluxDB::Rails::Logger
```
Comment (dmke) [inline lib/influxdb/rails/middleware/request_subscriber.rb]: The logger can be included in the base class as well.

File: spec/support/rails4/app.rb [hunk 4]
```diff
@@ -18,9 +18,9 @@
 
 class ApplicationController < ActionController::Base; end
 class WidgetsController < ApplicationController
-  def index
-    render nothing: true
-  end
+  prepend_view_path File.join(File.dirname(__FILE__), "..", "views")
```
Comment (dmke) [inline spec/support/rails4/app.rb]: `File.join(__dir__, "..", "views")`

File: spec/support/rails5/app.rb [hunk 5]
```diff
@@ -19,9 +19,9 @@
 
 class ApplicationController < ActionController::Base; end
 class WidgetsController < ApplicationController
-  def index
-    head 200
-  end
+  prepend_view_path File.join(File.dirname(__FILE__), "..", "views")
```
Comment (dmke) [inline spec/support/rails5/app.rb]: Same here.

File: spec/support/views/widgets/_item.html.erb [hunk 6]
```diff
@@ -0,0 +1 @@
+<div>item</div
```
Comment (dmke) [inline spec/support/views/widgets/_item.html.erb]: Please ensure that all files end in a newline.

File: spec/unit/middleware/render_subscriber_spec.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  describe ".call" do
+    let(:start_time)   { Time.at(1_517_567_368) }
+    let(:finish_time)  { Time.at(1_517_567_370) }
+    let(:series_name) { "series_name" }
+    let(:payload) { { identifier: "index.html", count: 43, cache_hits: 42 } }
+    let(:result) {
+      {
+        values:
+        {
```
Comment (dmke) [inline spec/unit/middleware/render_subscriber_spec.rb]: Please keep key and open brace on the same line.

**Review by dmke — CHANGES_REQUESTED** [review]
Apart from a few nitpicks, I like what I see.

**Comment by ChrisBr** [comment]
@dmke thanks for the review, implemented your suggestions.

**Comment by dmke** [comment]
I've changed a few bits (Rubocop was still producing output), but this is now merged.

**Pull request merged**